src/compiler/checker.ts

```diff
@@ -24325,7 +24325,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                             }
                             else if (middleLength === 1 && elementFlags[startLength] & ElementFlags.Rest) {
                                 // Middle of target is exactly one rest element. If middle of source is not empty, infer union of middle element types.
-                                const restType = isTupleType(source) ? getElementTypeOfSliceOfTupleType(source, startLength, endLength) : getTypeArguments(source)[0];

```

Comment:
When working on something else I noticed that within the alternate branch `source` became `never`. This meant that either the branch is redundant or that types are not quite right. Since the tests are passing without this branch, it's probably the former.

Reply:
There other redundant `isTupleType` checks: in the preceding `else if` block, and in two predicates in the first and last `if`s inside `if (middleLength === 2)`. Can you try removing those too?

Reading the code, the checks all look redundant since this `else` block has `isTupleType` proven from its preceding `if` predicate including `!isTupleType`.